In [1]:
#LOADING LIBRARIES
import pandas as pd
import numpy as np
import time
import random
import math
import warnings
warnings.filterwarnings('ignore')

In [2]:
#GIVEN GD FUNCTION
def gradient_descent(gradient,init_,learn_rate,n_iter=50,tol=1e-06):
    x=init_  #input
    for i in range(n_iter):
        delta = -learn_rate*gradient(x)
        if np.all(np.abs(delta)<=tol):
            break
        x+= delta
    return round(x*1000)/1000

## 1(i)

$$
\begin{array}
(i)~f(x)=x^{2}+3 x+4 \\
~~~~~f^{\prime}(x)=2 x+3 \\
(ii)~g(x)=x^{4}-3 x^{2}+2 x \\
~~~~~~g^{\prime}(x)=4 x^{3}-6 x+2
\end{array}
$$

In [3]:
num =  0.01        #increment value after each iteration for
slope1 = []          # collecting all slopes returned by gradient
initial_val = [-2,-1,0,1,2]    # my range fo inital values[-2,2]

In [4]:
for j in initial_val:
    lr_rate = 0.1
    for i in range(10):
        t = gradient_descent(lambda v: 2 * v + 3,j,lr_rate)
        slope1.append(t)
        lr_rate = num+0.01     

In [5]:
print('MINIMA for f(x) is :',min(slope1))

MINIMA for f(x) is : -1.565


1(ii)

In [6]:
num =  0.01        #increment value after each iteration for
slope2 = []          # collecting all slopes returned by gradient
initial_val = [-2,-1,0,1,2]    # my range fo inital values[-2,2]

In [7]:
for j in initial_val:
    lr_rate = 0.1
    for i in range(10):                  #performing with 10 learning rates
        t = gradient_descent(lambda v: 4 * (v**3) - (6*v) + 2,j,lr_rate)
        slope2.append(t)
        lr_rate = num+0.01         #updating learningrate   

In [8]:
#g(x) has two minimas
minima= set(slope2)

In [9]:
print('MINIMA for g(x) is :',minima)

MINIMA for g(x) is : {1.0, -1.366}


## 1b) Gradient function for LINEAR REGRESSION  =>  y = mx+b ,  x belongs to $R^{D}$

MEAN SQUARED ERROR LOSS FUNCTION
$$
\begin{aligned}
L(s) &=\frac{1}{n}\sum_{i=1}^{N}\left(y_{i}-\hat{y}_{i}\right)^{2} \\
&=\frac{1}{n}\sum_{i=1}^{N}\left[y_{i}-\left(m x_{i}+c\right)\right]^{2}
\end{aligned}
$$

DERIVATIVES OF SLOPE AND INTERCEPT FORM
$$
\begin{aligned}
\frac{\partial(s)}{\partial(m)}=\frac{2}{n} \cdot \sum_{i=1}^{N}\left(y_{i}-\hat{y}_{i}\right) \cdot\left(-x_{i}\right)$
\\
\frac{\partial(s)}{\partial(b)}=\frac{-2}{n} \cdot \sum_{i=1}^{N}\left(y_{i}-\hat{y}_{i}\right)$
\end{aligned}
$$

In [10]:
#gradient func that can work for multivaraible linear regression also
def gradient(x,y,m_curr,b_curr):
    y_predict = np.dot(x,m_curr) + b_curr
    inter_dr = -2 * np.mean(y-y_predict)
    coeff_dr = (-2)* np.dot(y-y_predict,x)
    g = [inter_dr,coeff_dr]
    return g

## 1c) GRADIENT DESCENT

In [11]:
#INPUT DATA
np.random.seed(0)
x=2.5 * np.random.rand(10000)+1.5
res = 1.5 *np.random.rand(10000)
y = 2+0.3 * x +res

In [12]:
#GRADIENT FUNCTION 
def gradient1gd(x,y,m_curr,b_curr):
    n= x.shape[0]         #number of rows

    y_predict = np.dot(x,m_curr) + b_curr    #PREDICTED Y VALUE
    
    inter_dr = -2 * np.mean(y-y_predict)                #INTERCEPT derivative
    coeff_dr = ((-2/n)* np.dot(y-y_predict,x))          #COEFFICIENTS derivative
    return np.array([coeff_dr,inter_dr])        
    

In [13]:
def gradient_descent(gradient1,init_,learn_rate,X_train,y_train,n_iter=1000,tol=1e-06):
    x  =init_               #coefficent,Intercepts are stored
    for i in range(n_iter):
        g = gradient1(X_train,y_train,x[0],x[1])      #calculating GRADIENT
        delta = -(learn_rate * g)                   #STEPSIZE
        if (np.all(np.abs(delta[1])<=tol)) & (np.all(np.abs(delta[0])<=tol)):   #IF BOTH allcoeff and intercept are less than 10^-6
            print('breaked at iteration num--',i)
            break
        x[0]+=delta[0]           #UPDATING CEOFFICIENTS
        x[1]+=delta[1]           #UPDATING INTEERCEPT
    res = [np.round(i*1000)/1000 for i in x]     #RETURING COEFF,INTERCEPT
    return res

In [14]:
start_time = time.time()
r1 =gradient_descent(gradient1gd,[1,0],0.1,x,y)    #(2,0.3) here w,b
print('INTERCEPT: {} , SLOPE: {}'.format(r1[0],r1[1]))
print('time taken is: ',time.time()-start_time)

breaked at iteration num-- 881
INTERCEPT: 0.304 , SLOPE: 2.732
time taken is:  0.10027909278869629


CALCULATING FOR DIFFERENT LEARNING RATES

In [15]:
lr_rate = [0.01,0.03,0.05,0.07,0.1]
for i in range(len(lr_rate)):
    start_time = time.time()
    r1 =gradient_descent(gradient1gd,[0,0],lr_rate[i],x,y)
    t = time.time()-start_time
    print('INTERCEPT: {} , SLOPE: {} ,with l_rate = {},time: {}'.format(r1[0],r1[1],lr_rate[i],t))
    
    

INTERCEPT: 0.555 , SLOPE: 2.001 ,with l_rate = 0.01,time: 0.11586499214172363
INTERCEPT: 0.328 , SLOPE: 2.662 ,with l_rate = 0.03,time: 0.131483793258667
INTERCEPT: 0.306 , SLOPE: 2.726 ,with l_rate = 0.05,time: 0.08461928367614746
INTERCEPT: 0.304 , SLOPE: 2.732 ,with l_rate = 0.07,time: 0.10024094581604004
breaked at iteration num-- 871
INTERCEPT: 0.304 , SLOPE: 2.732 ,with l_rate = 0.1,time: 0.06899499893188477


THIS DATASET HAS OPTIMAL LEARNING RATES = [0.07,0.1] for GRADIENT DESCENT FOR 1000 iterations 
(around 870th iteration it breaks)

## 1D) MINIBATCH  GRADIENT DESCENT

In [16]:
def gradient3(x,y,m_curr,b_curr):
    n= x.shape[0]    #NUMBER OF ROWS
    
    y_predict = np.dot(x,m_curr) + b_curr     #PREDICTED Y VALUE
    
    inter_dr = -2 * np.mean(y-y_predict)                         #INTERCEPT derivative
    coeff_dr = ((-2)* np.dot(y-y_predict,x))/n                   #COEFFICIENTS derivative
    return np.array([coeff_dr,inter_dr])

In [17]:
def MBgradient_descent(gradient,init_,learn_rate,X_train,y_train,batch_size,n_iter=1000,tol=1e-06):
    x  =init_
    inner_break = 0
    for i in range(n_iter):
        idx = random.sample(range(X_train.shape[0]),batch_size)       #SELECTING THE INDEX's OF A BATCH RANDOMLY
        g = gradient(X_train[idx],y_train[idx],x[0],x[1])                      #calculating GRADIENT by sending a batch
        delta = -(learn_rate * g)                                             #STEPSIZE
        if (np.all(np.abs(delta[1])<=tol)) & (np.all(np.abs(delta[0])<=tol)):
            break
        x[0]+=delta[0]                    #UPDATING CEOFFICIENTS
        x[1]+=delta[1]                     #UPDATING INTERCEPT
        
    res = [np.round(i*1000)/1000 for i in x]
    return res

In [18]:
start_time = time.time()
print(MBgradient_descent(gradient3,[1,0],0.1,x,y,800))
print('time taken is: ',time.time()-start_time)

[0.308, 2.73]
time taken is:  0.7197997570037842


CALCULATING FOR DIFFERENT LEARNING RATES AND DIFFERENT BATCHSIZES

In [19]:
lr_rate = [0.07,0.1]
batchsize= [100,200,300,400,500,600,700,800,900,1000]
for j in range(len(batchsize)):
    for i in range(len(lr_rate)):
        r1 =MBgradient_descent(gradient3,[1,0],lr_rate[i],x,y,batchsize[j])
        print('INTERCEPT: {} , SLOPE: {}  with l_rate: {}, batchsize: {}'.format(r1[0],r1[1],lr_rate[i],batchsize[j]))

INTERCEPT: 0.294 , SLOPE: 2.736  with l_rate: 0.07, batchsize: 100
INTERCEPT: 0.263 , SLOPE: 2.699  with l_rate: 0.1, batchsize: 100
INTERCEPT: 0.336 , SLOPE: 2.742  with l_rate: 0.07, batchsize: 200
INTERCEPT: 0.288 , SLOPE: 2.733  with l_rate: 0.1, batchsize: 200
INTERCEPT: 0.316 , SLOPE: 2.732  with l_rate: 0.07, batchsize: 300
INTERCEPT: 0.3 , SLOPE: 2.73  with l_rate: 0.1, batchsize: 300
INTERCEPT: 0.319 , SLOPE: 2.739  with l_rate: 0.07, batchsize: 400
INTERCEPT: 0.307 , SLOPE: 2.73  with l_rate: 0.1, batchsize: 400
INTERCEPT: 0.293 , SLOPE: 2.73  with l_rate: 0.07, batchsize: 500
INTERCEPT: 0.303 , SLOPE: 2.732  with l_rate: 0.1, batchsize: 500
INTERCEPT: 0.294 , SLOPE: 2.726  with l_rate: 0.07, batchsize: 600
INTERCEPT: 0.316 , SLOPE: 2.738  with l_rate: 0.1, batchsize: 600
INTERCEPT: 0.31 , SLOPE: 2.73  with l_rate: 0.07, batchsize: 700
INTERCEPT: 0.306 , SLOPE: 2.735  with l_rate: 0.1, batchsize: 700
INTERCEPT: 0.299 , SLOPE: 2.734  with l_rate: 0.07, batchsize: 800
INTERCEPT

In [20]:
def gradientsc(x,y,m_curr,b_curr):    
    y_predict = np.dot(x,m_curr) + b_curr
    
    inter_dr = -2 * (y-y_predict)
    coeff_dr = ((-2)* np.dot(y-y_predict,x))
    return np.array([coeff_dr,inter_dr])
    

In [21]:
def stochastic_gradient_descent(gradient,init_,learn_rate,X_train,y_train,n_iter=10000,tol=1e-06):
    x  =init_
    inner_break = 0
    for i in range(n_iter):
        idx = np.random.randint(0,X_train.shape[0]-1)             #SELECTING THE INDEX OF DATAPOINT RANDOMLY
        g = gradient(X_train[idx],y_train[idx],x[0],x[1])
        delta = -(learn_rate * g)                                #STEPSIZE

        if (np.all(np.abs(delta[1])<=tol)) & (np.all(np.abs(delta[0])<=tol)):
            print('breaked here at iteration--',i)
            inner_break =1
            break
        x[0]+=delta[0]                      #UPDATING CEOFFICIENTS
        x[1]+=delta[1]                      #UPDATING INTERCEPT
    
    res = [np.round(i*1000)/1000 for i in x]
    return res

In [22]:
start_time = time.time()
print(stochastic_gradient_descent(gradientsc,[1,0],0.006,x,y))           #0.3,2
print('time taken is: ',time.time()-start_time)

breaked here at iteration-- 2447
[0.398, 2.268]
time taken is:  0.07452154159545898


In [23]:
lr_rate = 0.004
for i in range(15):
    start_time = time.time()
    r1 =stochastic_gradient_descent(gradientsc,[1,0],lr_rate,x,y)
    timetaken = time.time() - start_time
    print('INTERCEPT: {} , SLOPE: {}, with l_rate = {},timetaken: {}'.format(r1[0],r1[1],lr_rate,timetaken))
    lr_rate += 0.001
    
    

    

INTERCEPT: 0.323 , SLOPE: 2.699, with l_rate = 0.004,timetaken: 0.29438328742980957
breaked here at iteration-- 9499
INTERCEPT: 0.277 , SLOPE: 2.753, with l_rate = 0.005,timetaken: 0.28528928756713867
INTERCEPT: 0.292 , SLOPE: 2.709, with l_rate = 0.006,timetaken: 0.2983286380767822
INTERCEPT: 0.216 , SLOPE: 2.716, with l_rate = 0.007,timetaken: 0.29424500465393066
breaked here at iteration-- 9427
INTERCEPT: 0.299 , SLOPE: 2.758, with l_rate = 0.008,timetaken: 0.2754690647125244
INTERCEPT: 0.308 , SLOPE: 2.63, with l_rate = 0.009000000000000001,timetaken: 0.3031482696533203
INTERCEPT: 0.27 , SLOPE: 2.706, with l_rate = 0.010000000000000002,timetaken: 0.29170775413513184
INTERCEPT: 0.241 , SLOPE: 2.694, with l_rate = 0.011000000000000003,timetaken: 0.29370570182800293
INTERCEPT: 0.29 , SLOPE: 2.723, with l_rate = 0.012000000000000004,timetaken: 0.2951226234436035
INTERCEPT: 0.258 , SLOPE: 2.707, with l_rate = 0.013000000000000005,timetaken: 0.28639912605285645
INTERCEPT: 0.313 , SLOPE: 

FROM THE ABOVE WE CAN SAY THAT SGD CONVERGES FASTER TO OPTIMAL VALUE THAN GRADIENTDESCENT and 
MINI BATCH GRADIENTDESCENT.,BUT AT THE COMPROMISE OF PREECISION

## performance comparision 

OPTIMAL GRADIENT DESCENT FOR 1000 epochs   (reuires around 900) for this data  

OPTIMAL LR_RATES(0.07,0.1)

In [24]:
gd=[0,0]
timegd=[0,0]
#CALCULATING TIMES
start_time = time.time()
gd[0] = gradient_descent(gradient1gd,[0,0],0.07,x,y)
timegd[0] = time.time() - start_time

start_time = time.time()
gd[1] = gradient_descent(gradient1gd,[0,0],0.1,x,y)
timegd[1] = time.time() - start_time

breaked at iteration num-- 871


#OPTIMAL STOCHASTIC GRADIENT DESCENT  REQUIRES  (i.e. around 8000 iterations only) and sometimes even lessthan 1000 iterations

#OPTIMAL LR_RATES are randomized and unpredicatble but for this data (0.005,0.01) are possible.

In [25]:
sgd = [0,0]
timesc=[0,0]
#CALCULATING TIMES
start_time = time.time()
sgd[0] = stochastic_gradient_descent(gradientsc,[1,0],0.005,x,y)
timesc[0] = time.time() - start_time

start_time = time.time()
sgd[1] = stochastic_gradient_descent(gradientsc,[1,0],0.01,x,y,n_iter=1000)
timesc[1] = time.time() - start_time

breaked here at iteration-- 3872


In [26]:
timegd,timesc

([0.11587190628051758, 0.10028791427612305],
 [0.19533634185791016, 0.032259464263916016])

CONCLUSION-- STOCASTIC CONVERGES FASTER TO OPTIMAL BUT DUE TO RANDOMNESS IT IS NOT ALWAYS GURANTEED

##  1E)OPTIMAL BATCH SIZE FOR MINIBATCH GRADIENT DESCENT

In [27]:
# FOR 1000 ITERATIONS
batch_sizes_mgd = [100,200,300,400,500,600,700,800,900,1000,5000]

lr_rate_mgd = [0.07,0.1]        #WITH OPTIMAL RATES OF MGD FOR THIS DATA
timemgd = []

for i in range(len(batch_sizes_mgd)):
    for j in range(len(lr_rate_mgd)):
        start_time = time.time()
        MBgradient_descent(gradient3,[1,0],lr_rate_mgd[j],x,y,batch_sizes_mgd[i],n_iter=1000)
        timemgd.append(time.time() - start_time)

In [28]:
least_time = min(timemgd)

ind = timemgd.index(least_time)
size_of_listbatch = len(batch_sizes_mgd)

In [29]:
print('OPTIMAL BATCH SIZE FOR THIS DATA >>',batch_sizes_mgd[math.floor(ind/2)])

OPTIMAL BATCH SIZE FOR THIS DATA >> 100


In [30]:
# FOR 100 ITERATIONS
timemgd = []

for i in range(len(batch_sizes_mgd)):
    for j in range(len(lr_rate_mgd)):
        start_time = time.time()
        MBgradient_descent(gradient3,[1,0],lr_rate_mgd[j],x,y,batch_sizes_mgd[i],n_iter=100)
        timemgd.append(time.time() - start_time)

In [31]:
least_time = min(timemgd)

ind = timemgd.index(least_time)
size_of_listbatch = len(batch_sizes_mgd)

In [32]:
print('OPTIMAL BATCH SIZE FOR THIS DATA >>',batch_sizes_mgd[math.floor(ind/2)])

OPTIMAL BATCH SIZE FOR THIS DATA >> 200


CONCLUSION--

-MINIBATCH GRADIENT DESCENT IS ALSO ASSOCIATED WITH RANDOMNESS

-- PERFORMANCE OF MGD DEPENDS ON BATCH SIZE AND NUMBER_OF_ iTERATIONS SIMULATANEOULSY

# Q2. Bayesian Network

![](Bayesian_Network.jpg)

(i) The probability that someone has both cold and a fever

\begin{aligned}
&P(\text {Cold and Fever})\\
&=P(C \cap F)\\
&=P(F \mid C) P(C)\\
&=0.307\times 0.02\\
&=0.00614\\
\end{aligned}

(ii) The  probability that  someone  who has  a  cough has  a  cold.

\begin{aligned}
&P(\text {Cold} \mid \text {Cough})=\frac{P(\text {Cold and Cough})}{P(\operatorname{Cough})}\\
&P({Cough})=P(\text {Lung}) P\left(Cold) P(\text {Cough} \mid \text {Lung and Cold})+\right.P(\neg Lung) P(\text {Cold}) P(Cough \mid {\neg Lung \space and \space Cold)}+P(\neg Cold) P(\text {Lung}) P(Cough \mid {\neg Cold\space and \space Lung)}+P(\neg Cold) P(\neg {Lung}) P(Cough \mid {\neg Lung \space and \space \neg Cold)}.\\
&=P(\text { Lung })\times 0.02 \times 0.7525 +\\
&~~~~~ P{(\neg Lung) }\times 0.02 \times0.505+\\
&~~~~~ P\text{(Lung) }\times 0.98 \times0.505+\\
&~~~~~ P{(\neg Lung) }\times 0.98 \times 0.01\quad-(1)\\
&P(\text {Lung})=P(\text {Lung} \mid \text {Smokes)}\times P(Smokes)+P(\text {Lung} \mid {\neg Smokes)}\times P(\neg Smokes)\\
&~~~~~~~~~~~~~~~=0.1009 \times 0.2+0.001 \times 0.8\\
&~~~~~~~~~~~~~~~=0.02098  \quad-(2)\\
&P(\neg Lung)=P(\neg Lung \mid \text {Smokes}) P(\text {Smokes})+P { (\neg Lung) }\mid {\neg Smokes) } P({ \neg Smokes) }\\
&~~~~~~~~~~~~~~~~=0.8991 \times 0.2+0.999 \times 0.8\\
&~~~~~~~~~~~~~~~~=0.97902\quad-(3)\\
&Putting ~ (2) ~ and ~ (3) ~ in ~ (1)\\
&P(Cough) = 0.02098\times 0.02 \times 0.7525 + 0.97902\times 0.02 \times0.505+ 0.02098\times 0.98 \times0.505+0.97902\times 0.98 \times 0.01\\
&P(Cough)= 0.030181 \\ 
&Now, \\
&{P(\text {Cold and Cough})} =P(Cough \mid Cold) P(Cold) = P(Cough \mid Cold) \times 0.02 \\
&P(Cough \mid Cold) = P(Cough \mid Cold,Lung) \times P(Lung) + P(Cough \mid Cold, \neg Lung) \times P(\neg Lung)\\
& = 0.7525\times0.02098 + 0.505\times0.97902 \\
& P(Cough \mid Cold) = 0.5102\\
&{P(\text {Cold and Cough})} = 0.5102 \times 0.02 = 0.010204 \\
&Finally, \\
&Using \space P(Cold \space and  \space Cough) \space and \space P(Cough),\\
&P(\text {Cold} \mid \text {Cough}) = 0.010204 \div 0.030181 = 0.3381\\                                                                                                        
\end{aligned}

# Q3. Derive the MLE for the parameters of a k-sided multinomial distribution

Sol: Joint probability of Multinomial distribution is the Likelihood given as $L(P)$
$$
L(p)=\frac{n !}{x_{1} ! x_{2} ! x_{3} ! \ldots x_{k} !} p_{1}^{x_{1}} p_{2}^{x_{2}} p_{3}^{x_{3}} \ldots p_{k}^{x_{k}}
$$
Let $\theta:\left\{\theta_{1}, \theta_{2}, \ldots \theta_{k}\right\}$ such that $\sum_{i=1}^{k} \theta_{k}=1$ 
The log likelihood is,
$$
l(P)=\ln \left(\prod_{i=1}^{k} \theta_{i}^{N_{i}}\right) ;  \begin{gathered}
\space N_{i} \text { be the number of times face 'i'}
\text { appeared }
\end{gathered}
$$
Now, 
$\space \theta^{M L E}=\operatorname{argmax}\left(\ln \prod_{i=1}^{k} \theta_{i}^{N i}\right)\\$
In multinomial last term is not a random variable but a result of all the other random variables,
$\Rightarrow \theta^{M L E}=\operatorname{argmax}\ln \left[\left(\prod_{i=1}^{k-1} \theta_{i}^{N i}\right)\left(1-\sum_{i=1}^{k-1} \theta_{i}\right)^{N_{k}}\right]$
$=\operatorname{argmax}\left[\sum_{i=1}^{k-1}\right.$  $\left.N_{i} \ln \theta_{i}+N_{k} \ln \left(1-\sum_{i=1}^{k-1} \theta_{i}\right)\right]$
$=\frac{\partial}{\partial \theta_{i}}\left[\sum_{i=1}^{k-1} N_{i} \ln \theta_{i}+N_{k} \ln \left(1-\sum_{i=1}^{k-1} \theta_{i}\right)\right]$
$\Rightarrow  \frac{N_{i}}{\theta_{i}}+\frac{N_{k}}{1-\sum_{i=1}^{N-1} \theta_{i}}(-1)=0$
$\Rightarrow \theta_{i}=\frac{N_{i}}{N_{k}} \theta_{k}$
We know that, $\theta_{1}+\theta_{2}+\cdots+\theta_{k}=1$
$$
\begin{aligned}
&\Rightarrow\left[\frac{N_{1}}{N_{k}} \theta_{k}+\frac{N_{2}}{N_{k}} \theta_{k}+\cdots \frac{N_{k}}{N_{k}} \theta_{k}\right]=1 \\
&\Rightarrow \quad \theta_{k}=\frac{N_{k}}{\sum_{i=1}^{k} N_{i}}=\frac{N_{k}}{N} \\
&\therefore \quad \theta_{k}=\frac{N_{k}}{N} ; where \space N=N_{1}+N_{2}+\cdots+N_{k}
\end{aligned}
$$
The probability distribution that maximizes the likelihood of observing the data is $\quad P=\left(\frac{N_{1}}{N}, \frac{N_{2}}{N},\cdots,\frac{N_{k}}{N}\right)$

#FINISHED